### PR TITLE
Issue #133 file locations for distances matrix()

### PR DESCRIPTION
**Pull request by dsmith-unbc**
Updated `distanceMatrix()` to accept a new condition 'file_path'. This will solve issue [#133] where if the 'spatial.csv' is not in the main working directory the function will error. New checks have also been added for 'file_path in' `distanceMatrix()` plus the documentation for `distanceMatrix()` has been updated to reflect 'file_path' being added.

### CODE REVIEW

**Comment by dsmith-unbc** [comment]
Notes based on discussion with Hugo:

If distancesMatrix() is called without starters, have function look for spatial.csv in current working directory (this is the current functionality). 

Utilize the loadSpatial() function being called before distancesMatrix() for other processes. This would have the existing path of spatial.csv already denoted in it. Use the output of loadSpatial() as a input for distancesMatrix() which would be put in the starters argument. This would bypass the need for an additional argument while still providing the information needed to run distancesMatrix(). 

Starters and targets in distancesMatrix() define endpoints used to calculate distances. Normally, these two are the same values (as we want to calculate the distances between every pair of points). 

Currently, if actel = TRUE, distancesMatrix() will ignore the starters argument. However, distancesMatrix() should inherently always try to optimize the output for actel. 

Summary: Relax the distancesMatrix() function so it accepts starters regardless of actel being TRUE or not. Allow the user to supply the output of loadSpatial() to starters so the spatial.csv can be accessed. Keep the default functionality of distanceMatrix() the same where if no starters are provided, distancesMatrix() looks for a spatial.csv in the current working directory. 

@hugomflavio Please give this a read through and provide feedback if something requires rectifying.

**Comment by hugomflavio** [comment]
Looks good. One comment:

> Currently, if actel = TRUE, distancesMatrix() will ignore the starters argument. However, distancesMatrix() should inherently always try to optimize the output for actel.

I'd rephrase this thought to: However, distancesMatrix() should try to optimize the output for actel if `actel = TRUE` even if starters/targets arguments have been provided by the user.

Had a quick look at distancesMatrix() to see if anything else could go wrong. It seems it already has fail-safes for if id.col does not exist. It still makes sense for force id.col to be "Standard.name" when actel = TRUE, as that is what actel will look for. I suppose distancesMatrix() should crash if 1) actel = TRUE and 2) starters/targets were supplied but do not contain a column called "Standard.name".

**Comment by dsmith-unbc** [comment]
First push to address issue from a loadSpatial() approach. Function will now accept loadSpatial() output for starters and if no starters or targets are specified, then function searches and grabs 'spatial.csv' from working directory as done previously. 

The function will now allow starters and targets to be specified when actel == TRUE. 

If starters are applied and targets are left blank, starters will be used as targets. 

Added new code to the end allowing the user to specify the folder where the 'distances.csv' should be saved to. Or the user can have it left default and the 'distances.csv' will be saved to the working directory. 

Relevant documentation has been updated as well as news. 

The distancesMatrix() examples in the help documentation did seem confusing we should considering altering that to be more intuitive of how the function works right now.

File: R/distances.R [hunk 1]
```diff
@@ -419,41 +430,104 @@ distancesMatrix <- function(t.layer, starters = NULL, targets = starters,
...
   missing.packages <- sapply(aux, function(x) x == 1)
   if (any(missing.packages)) {
-    stop(paste0("This function requires packages '", paste(c("raster", "gdistance", "sp", "terra")[missing.packages], collapse = "', '"),
-      "' to operate. Please install ", ifelse(sum(missing.packages) > 1, "them", "it"), " before proceeding.\n"), call. = FALSE)
+    stop(paste0("This function requires packages '", 
+                paste(c("raster", "gdistance", "sp", "terra")[missing.packages], 
+                      collapse = "', '"),
+      "' to operate. Please install ", 
+      ifelse(sum(missing.packages) > 1, "them", "it"), 
+      " before proceeding.\n"), call. = FALSE)
```
Comment (hugomflavio) [inline R/distances.R:438]:
```suggestion
    stop("This function requires packages '", 
         paste(c("raster", "gdistance", "sp", "terra")[missing.packages], 
               collapse = "', '"),
         "' to operate. Please install ", 
         ifelse(sum(missing.packages) > 1, "them", "it"), 
         " before proceeding.\n", call. = FALSE)
```

Taking the chance to remove an unnecessary paste0()

File: R/distances.R [hunk 2]
```diff
@@ -419,41 +430,104 @@ distancesMatrix <- function(t.layer, starters = NULL, targets = starters,
...
-      warning("starters' or 'targets' were set but will be ignored because 'actel' is set to TRUE. Set 'actel' to FALSE to use the 'starters' and 'targets' arguments.", call. = FALSE, immediate. = TRUE)
-    starters <- targets <- loadSpatial()
-    if (!is.null(id.col))
-      warning("id.col' was set but will be ignored because 'actel' is set to TRUE. Set 'actel' to FALSE to use the 'id.col' argument.", call. = FALSE, immediate. = TRUE)
+    
+    if(!is.null(starters) | !is.null(targets)) {
+      # distancesMatrix() crashes if actel = TRUE and starters/targets were 
+      # supplied but do not contain a column called "Standard.name"
+      if(!is.null(starters) & !any(grepl(c("Standard.name"), 
+                                         colnames(starters)))) {
```
Comment (hugomflavio) [inline R/distances.R:459]:
```suggestion
      if(!is.null(starters) && !("Standard.name" %in% colnames(starters))) {
```
`&&` instead of & so colnames doesn't return NULL if starters is NULL.
`%in%` allows keeping in one line

File: R/distances.R [hunk 3]
```diff
@@ -419,41 +430,104 @@ distancesMatrix <- function(t.layer, starters = NULL, targets = starters,
...
+        event("'starters' is missing column called 'Standard.name'",
+              type = "stop")
```
Comment (hugomflavio) [inline R/distances.R:461]:
```suggestion
        stop("'starters' is missing column called 'Standard.name'",
             call. = FALSE)
```

distancesMatrix operates outside the actel analyses, so it doesn't need to use event() - at least for now. Best to keep it consistent until we decide we need to change everything to events.

File: R/distances.R [hunk 4]
```diff
@@ -419,41 +430,104 @@ distancesMatrix <- function(t.layer, starters = NULL, targets = starters,
...
+      }
+      
+      if(!is.null(targets) & !any(grepl(c("Standard.name"), 
+                                         colnames(targets)))) {
+        event("'targets' is missing column called 'Standard.name'",
+              type = "stop")
```
Comment (hugomflavio) [inline R/distances.R:467]:
```suggestion
        stop("'targets' is missing column called 'Standard.name'",
             call. = FALSE)
```

same as above

File: R/distances.R [hunk 5]
```diff
@@ -419,41 +430,104 @@ distancesMatrix <- function(t.layer, starters = NULL, targets = starters,
...
+      
+        targets <- starters
+        }
+    
+    if (is.null(starters) & is.null(targets)) {
+        warning("'starters' and 'targets' were not set while 'actel' is set to",
+              " TRUE. Creating 'starters' and 'targets' from 'spatial.csv' ",
+              " in working directory." , call. = FALSE, immediate. = TRUE)
+        starters <- targets <- loadSpatial() 
+        }
```
Comment (hugomflavio) [inline R/distances.R:490]: do we need these three warnings? I think the first two could go away; it's just the function doing what it was ordered to do. The last one could be downgraded to a message. Something perhaps simpler e.g.

```
        message("'starters' and 'targets' were not set.",
                " Creating 'starters' and 'targets' from 'spatial.csv' ",
                " in working directory.")
```

What do you think?

File: R/distances.R [hunk 6]
```diff
@@ -419,41 +430,104 @@ distancesMatrix <- function(t.layer, starters = NULL, targets = starters,
...
+                coord.y, "' in 'starters'."), call. = FALSE)
+    }
+  if (is.na(match(coord.x, colnames(targets)))) {
+    stop(paste0("Could not find a column '", 
+                coord.x, "' in 'targets'."), call. = FALSE)
+    }
+  if (is.na(match(coord.y, colnames(targets)))) {
+    stop(paste0("Could not find a column '", 
+                coord.y, "' in 'targets'."), call. = FALSE)
+    }
```
Comment (hugomflavio) [inline R/distances.R:530]:
```suggestion
  if (is.na(match(coord.x, colnames(starters)))) {
    stop(paste0("Could not find a column '", 
                coord.x, "' in 'starters'."), call. = FALSE)
  }
  if (is.na(match(coord.y, colnames(starters)))) {
    stop(paste0("Could not find a column '", 
                coord.y, "' in 'starters'."), call. = FALSE)
  }
  if (is.na(match(coord.x, colnames(targets)))) {
    stop(paste0("Could not find a column '", 
                coord.x, "' in 'targets'."), call. = FALSE)
    }
  if (is.na(match(coord.y, colnames(targets)))) {
    stop(paste0("Could not find a column '", 
                coord.y, "' in 'targets'."), call. = FALSE)
  }
```

File: R/distances.R [hunk 7]
```diff
@@ -493,45 +575,108 @@ distancesMatrix <- function(t.layer, starters = NULL, targets = starters,
 
 
   #### Create starters and targets spatial dataframes
-  sp::coordinates(starters) <- ~ longitude + latitude # converts the file to a spatialPoints object
-  raster::crs(starters) <- raster::crs(t.layer) # sets the crs
-
-  sp::coordinates(targets) <- ~ longitude + latitude # converts the file to a spatialPoints object
+    # converts the file to a spatialPoints object
+  sp::coordinates(starters) <- ~ longitude + latitude
```
Comment (hugomflavio) [inline R/distances.R:579]:
```suggestion
  # converts the file to a spatialPoints object
  sp::coordinates(starters) <- ~ longitude + latitude 
```

File: R/distances.R [hunk 8]
```diff
@@ -493,45 +575,108 @@ distancesMatrix <- function(t.layer, starters = NULL, targets = starters,
...
+            " in the shape file, consider applying a 'buffer' when calculating", 
+            " the transition layer. This will artificially add water space", 
+            " around the shape file.", call. = FALSE)
     dist.mat[dist.mat == Inf] <- NA
   }
 
-  if (row.rename)
+  if (row.rename) {
     rownames(dist.mat) <- outputRows
+    }
```
Comment (hugomflavio) [inline R/distances.R:611]:
```suggestion
  }
```

File: R/distances.R [hunk 9]
```diff
@@ -493,45 +575,108 @@ distancesMatrix <- function(t.layer, starters = NULL, targets = starters,
...
 
-  if (col.rename)
+  if (col.rename) {
     colnames(dist.mat) <- outputCols
+    }
```
Comment (hugomflavio) [inline R/distances.R:615]:
```suggestion
  }
```

File: R/distances.R [hunk 10]
```diff
@@ -493,45 +575,108 @@ distancesMatrix <- function(t.layer, starters = NULL, targets = starters,
...
 
   if (interactive() & actel) { # nocov start
-    decision <- userInput("Would you like to save an actel-compatible distances matrix as 'distances.csv' in the current work directory?(y/n) ",
+    decision <- userInput(paste0("Would you like to save an actel-compatible", 
+                                 " distances matrix as 'distances.csv'?(y/n)"),
+                          choices = c("y", "n"))
+    if (decision == "y") {
+       dir_decision <- userInput(paste0("Would you like 'distances.csv' saved in",
+                                    " the current working directory?(y/n)"),
                           choices = c("y", "n"))
```
Comment (hugomflavio) [inline R/distances.R:624]:
```suggestion
       dir_decision <- userInput(paste0("Would you like 'distances.csv'",
                                        " saved in the current working",
                                        " directory?(y/n)"),
                                 choices = c("y", "n"))
```

File: R/distances.R [hunk 11]
```diff
@@ -493,45 +575,108 @@ distancesMatrix <- function(t.layer, starters = NULL, targets = starters,
...
+       if (dir_decision == "n") {
+         
+         target_dir <- ""
+
+         while(!dir.exists(target_dir)) {
```
Comment (hugomflavio) [inline R/distances.R:629]:
```suggestion
         while (!dir.exists(target_dir)) {
```

File: R/distances.R [hunk 12]
```diff
@@ -493,45 +575,108 @@ distancesMatrix <- function(t.layer, starters = NULL, targets = starters,
...
+           target_dir <- readline(paste0("Specify the folder to save 'distances.csv'",
+                                         " to. Hit enter to skip."))
```
Comment (hugomflavio) [inline R/distances.R:631]:
```suggestion
           target_dir <- readline(paste0("Specify the folder to save",
                                         " 'distances.csv' to.",
                                         " Hit enter to skip."))
```

File: R/distances.R [hunk 13]
```diff
@@ -493,45 +575,108 @@ distancesMatrix <- function(t.layer, starters = NULL, targets = starters,
...
+           if(nchar(target_dir) == 0) {
```
Comment (hugomflavio) [inline R/distances.R:632]:
```suggestion
           if (nchar(target_dir) == 0) {
```

File: R/distances.R [hunk 14]
```diff
@@ -493,45 +575,108 @@ distancesMatrix <- function(t.layer, starters = NULL, targets = starters,
...
+             break
+           } 
+           
+           if(!dir.exists(target_dir)) {
+             dir_decision <- userInput(paste0("Directory specified not found.",
+                                              " Do you want to continue?(y/n)"),
+                       choices = c("y", "n"))
+             if(dir_decision == "n") {
+               dir_decision <- "y"
+               break
```
Comment (hugomflavio) [inline R/distances.R:643]:
```suggestion
           if (!dir.exists(target_dir)) {
             dir_decision <- userInput(paste0("Directory specified not found.",
                                              " Do you want to continue?(y/n)"),
                                       choices = c("y", "n"))
             if (dir_decision == "n") {
               dir_decision <- "y"
               break
```

File: R/distances.R [hunk 15]
```diff
@@ -493,45 +575,108 @@ distancesMatrix <- function(t.layer, starters = NULL, targets = starters,
...
     if (decision == "y") {
-      if (file.exists('distances.csv')) {
-        warning("A file 'distances.csv' is already present in the current directory.", call. = FALSE, immediate. = TRUE)
-        decision <- userInput("Continuing will overwrite this file. Would you like to continue?(y/n) ", choices = c("y", "n"))
+      
+      if(dir_decision == "n") {
+        
+        # If a different directory has been specified, a new path is created  
+        # where 'distances.csv' will be saved
+        path <- paste(target_dir,"distances.csv", sep = "/")
```
Comment (hugomflavio) [inline R/distances.R:656]:
```suggestion
        path <- paste0(target_dir, "/distances.csv")
```

File: R/distances.R [hunk 16]
```diff
@@ -493,45 +575,108 @@ distancesMatrix <- function(t.layer, starters = NULL, targets = starters,
...
+        
+        } else {
+        # If no 'path' specified, then 'distances.csv' is saved to the 
+        # current working directory
+        
+        path <- paste("distances.csv", sep = "/")
+        
+        }
```
Comment (hugomflavio) [inline R/distances.R:664]:
```suggestion
      } else {
        # If no 'path' specified, then 'distances.csv' is saved to the 
        # current working directory
        
        path <- paste("distances.csv", sep = "/")
        
      }
```

**Review by hugomflavio — CHANGES_REQUESTED** [review]

File: R/distances.R (hunk 5 above)
Comment (dsmith-unbc) [reply R/distances.R:489]: @hugomflavio  

My objective was to add more user feedback as to how the function was operating in different situations. While they do not provide critical information to the user, they do provide some insight into how the function is using the data they provided to create the distances matrix. 

I do agree they are not "warnings" and could be messages instead. 

What are your thoughts? 

I can defiantly remove the first two and only retain the last (as a message) if you'd prefer it that way. It really depends how much feedback you want to give to the user about the function's operations.

Comment (hugomflavio) [reply R/distances.R:489]: right, makes sense, since it is copying one argument to another. Can we make them messages instead and shorten by removing the actel = TRUE part? By this logic, this part of the code should also have a message, no?

image

File: R/distances.R [hunk 17]
```diff
@@ -455,16 +455,15 @@ distancesMatrix <- function(t.layer, starters = NULL, targets = NULL,
...
-                                         colnames(starters)))) {
-        event("'starters' is missing column called 'Standard.name'",
-              type = "stop")
+      if(!is.null(starters) && !("Standard.name" %in% colnames(starters))) {
+        stop("'starters' is missing column called 'Standard.name'",
+             call. = FALSE)
       }
       
       if(!is.null(targets) & !any(grepl(c("Standard.name"), 
                                          colnames(targets)))) {
```
Comment (hugomflavio) [inline R/distances.R:464]: Just realized we should change this one too

```suggestion
      if(!is.null(targets) && !("Standard.name" %in% colnames(targets))) {
```

File: R/distances.R (hunk 5 above)
Comment (dsmith-unbc) [reply R/distances.R:489]: Great point Hugo. I will add a message there and make those revisions now.
